### PR TITLE
Fix download with Token Authentication from funk.net (Nexx Player)

### LINKED ISSUE #24309: Download from funk.net fails with HTTP Error 403: Forbidden / HTTP Error 400: Bad Request
## Checklist

- [x] I'm reporting a broken site support
- [x] I've verified that I'm running youtube-dl version **2020.03.08**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that all URLs and arguments with special characters are properly quoted or escaped
- [x] I've searched the bugtracker for similar issues including closed ones

## Verbose log

```
$ youtube-dl --verbose  https://www.funk.net/channel/doctor-who-1164/journeys-end-ov-515545/doctor-who-staffel-4-ov-2473

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'https://www.funk.net/channel/doctor-who-1164/journeys-end-ov-515545/doctor-who-staffel-4-ov-2473']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2020.03.08
[debug] Python version 2.7.12 (CPython) - Linux-4.15.0-88-generic-x86_64-with-LinuxMint-18.1-serena
[debug] exe versions: ffmpeg 2.8.15, ffprobe 2.8.15, rtmpdump 2.4
[debug] Proxy map: {}
[Funk] 515545: Downloading JSON metadata
[Nexx] 515545: Downloading JSON metadata
[Nexx] 515545: Downloading m3u8 information
WARNING: Failed to download m3u8 information: HTTP Error 403: Forbidden
[Nexx] 515545: Downloading MPD manifest
WARNING: Failed to download MPD manifest: HTTP Error 403: Forbidden
[Nexx] 515545: Downloading ISM manifest
WARNING: Failed to download ISM manifest: HTTP Error 403: Forbidden
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on u'http://nx-d09.akamaized.net/06be6b49-c771-4f80-9ceb-a658ebc5e751/515545_src_1920x1080_6000.mp4'
ERROR: unable to download video data: HTTP Error 400: Bad Request
Traceback (most recent call last):
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 1926, in process_info
    success = dl(filename, info_dict)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 1865, in dl
    return fd.download(name, info)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/downloader/common.py", line 366, in download
    return self.real_download(filename, info_dict)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/downloader/http.py", line 341, in real_download
    establish_connection()
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/downloader/http.py", line 109, in establish_connection
    ctx.data = self.ydl.urlopen(request)
  File "/home/yee/Develop/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 2238, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
HTTPError: HTTP Error 400: Bad Request
```

## Description

The video https://www.funk.net/channel/doctor-who-1164/journeys-end-ov-515545/doctor-who-staffel-4-ov-2473 plays in a browser (Firefox) without problems, but `youtube-dl` fails to download m3u8 information, MPD & ISM manifests with *HTTP Error 403: Forbidden*.

About 3 weeks ago I didn't experience any problems downloading videos from this site.  Additionally, the problem only seems to occur when the location is **nx-d09**.akamaized.net like
```
[debug] Invoking downloader on u'http://nx-d09.akamaized.net/06be6b49-c771-4f80-9ceb-a658ebc5e751/515545_src_1920x1080_6000.mp4'
ERROR: unable to download video data: HTTP Error 400: Bad Request
```

Other videos, e.g. https://www.funk.net/channel/true-demon-12035/camping-im-wald-nimmt-fuerchterliches-ende-01-1616275/true-demon-alle-folgen-4971 from a different location **funk-01**.akamaized.net can still be downloaded:
```
$ youtube-dl --verbose https://www.funk.net/channel/true-demon-12035/camping-im-wald-nimmt-fuerchterliches-ende-01-1616275/true-demon-alle-folgen-4971

[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'https://www.funk.net/channel/true-demon-12035/camping-im-wald-nimmt-fuerchterliches-ende-01-1616275/true-demon-alle-folgen-4971']
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2020.03.08
[debug] Python version 2.7.12 (CPython) - Linux-4.15.0-88-generic-x86_64-with-LinuxMint-18.1-serena
[debug] exe versions: ffmpeg 2.8.15, ffprobe 2.8.15, rtmpdump 2.4
[debug] Proxy map: {}
[Funk] 1616275: Downloading JSON metadata
[Nexx] 1616275: Downloading JSON metadata
[Nexx] 1616275: Downloading m3u8 information
[Nexx] 1616275: Downloading MPD manifest
[Nexx] 1616275: Downloading ISM manifest
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on u'http://funk-01.akamaized.net/098fec4a-c9a8-446e-bcb0-d4ae19f2612c/1616275_src.ism/QualityLevels(5978504)/Manifest(video,format=m3u8-aapl)'
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 120
[download] Destination: Camping im Wald nimmt fürchterliches Ende #01-1616275.fazure-hls-6257.mp4
[download]   2.9% of ~567.00MiB at  5.55MiB/s ETA 01:46^C
ERROR: Interrupted by user
```

Thank you for your work.

### PR DESCRIPTION
**Pull request by yeeeha**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information
Fixes #24309 -- please see there for details.

Thank you for your work.  :)

### CODE REVIEW

File: youtube_dl/extractor/nexx.py [hunk 1]
```diff
@@ -247,20 +247,30 @@ def get_cdn_shield_base(shield_type='', static=False):
...
+            azure_manifest_url_DASH = azure_manifest_url + '?hdnts=%s' % protection_token_DASH
+        else:
+            azure_manifest_url_DASH = azure_manifest_url
+
+        protection_token_HLS = try_get(
+            video, lambda x: x['protectiondata']['tokenHLS'], compat_str)
+        if protection_token_HLS:
+            azure_manifest_url_HLS = azure_manifest_url + '?hdnts=%s' % protection_token_HLS
+        else:
+            azure_manifest_url_HLS = azure_manifest_url
```
Comment (dstftw) [inline youtube_dl/extractor/nexx.py]: DRY.

File: youtube_dl/extractor/nexx.py [hunk 2]
```diff
@@ -313,8 +323,16 @@ def find_video(result):
             if result:
                 video = find_video(result)
 
+        if video:
+            # search for protection tokens -- if they're present we have to go
+            # through the procedure below...
+            tokenDASH = try_get(
+                video, lambda x: x['protectiondata']['tokenDASH'], compat_str)
+            tokenHLS = try_get(
+                video, lambda x: x['protectiondata']['tokenHLS'], compat_str)
```
Comment (dstftw) [inline youtube_dl/extractor/nexx.py]: DRY.

Comment (yeeeha) [reply youtube_dl/extractor/nexx.py:332]: Both keys are used as indicators that the JSON metadata has to be re-fetched from a different location.  The values themselves are invalid.  They must be obtained from the *new* JSON metadata (s. above code).

File: youtube_dl/extractor/nexx.py (hunk 1 above)
Comment (yeeeha) [reply youtube_dl/extractor/nexx.py:263]: I can shorten it a bit...

File: youtube_dl/extractor/nexx.py (hunk 2 above)
Comment (dstftw) [reply youtube_dl/extractor/nexx.py:327]: Again: you must eliminate code duplication. In both places.

Comment (yeeeha) [reply youtube_dl/extractor/nexx.py:327]: You have noticed my last two commits?

**Comment by yeeeha** [comment]
@dstftw: Please review again.  Code duplication removed, the rest is symmetry.

**Comment by dstftw** [comment]
Nothing changed.

**Comment by yeeeha** [comment]
I'm talking about commits 264f0d5 and e125255.  Where do you see code duplication? IMHO all of it is necessary.

**Comment by dstftw** [comment]
Do you understand what is code duplication at all?

**Comment by yeeeha** [comment]
Very obviously we have quite different opinions about code duplication.  Show me the duplicate lines then.

**Comment by dstftw** [comment]
There can't be different opinions on basic programming principles. I've already shown.

**Comment by yeeeha** [comment]
As you insist on being mysterious I can only guess what's bothering you.
If you refer to the duplicate
```python
            tokenDASH = try_get(
                video, lambda x: x['protectiondata']['tokenDASH'], compat_str)
            tokenHLS = try_get(
                video, lambda x: x['protectiondata']['tokenHLS'], compat_str)
```
in lines 251-254 and lines 324-327 I already pointed out that re-fetching the values from a different JSON file is absolutely necessary.  It's the essence of the fix.
If you mean something else please tell me because I've got no idea what you mean.

**Comment by dstftw** [comment]
You don't understand what code duplication is. Bother to at least read wikipedia on that.

**Comment by yeeeha** [comment]
Seriously, is this satisfying you?
You can reject my pull request.  I'm fed up and I won't spend any more work on this.

**Comment by dstftw** [comment]
Satisfying me, huh? I don't care, really.
This will be accepted once these obvious review points are fixed. For now you are stubbornly refusing to do so for some reason.

**Comment by github-kh2019** [comment]
As I a have an interest in the fix I have read this exchange between @dstftw  and @yeeeha. And yes, I went to read the wikipedia entry [Duplicate code](https://en.wikipedia.org/wiki/Duplicate_code). After reading that entry, I must disagree: 
> There can't be different opinions on basic programming principles.

There CAN be different opinions about such loosely defined terms. The Wikipedia entry itself says
> A minimum requirement is **usually** applied to the quantity of code that must appear in a sequence for it to be **considered** duplicate

(emphasis is mine). A "**consideration**" can have different outcomes for different people.

I must admit, I also have a hard time to recognize the code duplication mentioned here. @dstftw , could you please explain what the duplication really is, in your opinion?

**Comment by FliegendeWurst** [comment]
The lookup `x['protectiondata']['tokenDASH']` (same for `tokenHLS`) is done twice.

**Pull request closed without merging**